File: Sequetial Search Q validation/functions/Maximization.R
###############################################################################
# Maximization:                                                               #
#                                                                             #
# Estimate the guessing and slipping parameters for an item                   #
#                                                                             #
# Input:                                                                      #
# (1) g        : guessing parameter for all items (n by 1)                    #
# (2) item     : the item to be estimated                                     #
# (3) s        : slipping parameter for all items (n by 1)                    #
# (4) Q        : Q matrix                                                     #
# (5) response : response matrix                                              #
#                                                                             #
# Output:                                                                     #
# (1) g_est  : estimated g value                                              # 
# (2) s_est  : estimated s value                                              # 
###############################################################################

Maximization <- function(g,item,s,Q,response) {
  
  alpha <- AlphaPermute(ncol(Q))
  post  <- Post_alpha(g,Q,response,s)  # posterior probability for each examinee, each pattern
  eta   <- Eta(alpha,Q)[,item]
  L     <- rowSums(post*rep(1,nrow(response)) %*% t(eta))
  
  n0 <- nrow(response)-sum(L)
  r0 <- sum(response[,item]*(1-L))
  n1 <- sum(L)
  r1 <- sum(response[,item]*L)
    
  g_est <- r0 / n0
  s_est <- (n1-r1) / n1
   
  return(list(g_est=g_est,s_est=s_est))
}

File: CDCAT_ONLINE/Maximization.m
function [s_est,g_est] = Maximization(s,g,item,Q,sparse)
%%% s g are column vectors
%%% item is the item number to be calibrated
%%% Q is the Q matrix
%%% sparse is the sparse matrix 
%%% s_est and g_est are column vectors

persons = find(sparse(:,item)~=9);
alpha_permute = AlphaPermute(Q);

post = post_alpha(s,g,Q,sparse);
post = post(persons,:);

eta = Eta(alpha_permute,Q);
eta = eta(:,item);
 
L = post * eta;

n0 = persons - sum(L);
r0 = sum(sparse(persons,item).*(1-L));
n1 = sum(L);
r1 = sum(response(persons,item).*L);

g_est = r0 / n0;
s_est = (n1-r1) / n1;

